Dockerfile: pin criu version and build from source

crazy-max opened the pull request:
related to #43529 

**- What I did**

pin criu for better reproducibility and build from source.
will solve issues we have with flakiness using the opensuse mirror
but also being able to build on other platforms:
* https://github.com/moby/moby/pull/44034#issuecomment-1232864547
* https://github.com/moby/moby/pull/34425#issuecomment-966986752
* https://github.com/moby/moby/pull/43680#issuecomment-1145715984
* https://github.com/moby/moby/pull/43681#issuecomment-1145155540
* https://github.com/moby/moby/pull/43682#issuecomment-1145116748
* https://github.com/moby/moby/pull/43130#issuecomment-1017243796

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

thaJeztah commented on Dockerfile:
```diff
@@ -16,15 +17,37 @@ RUN sed -ri "s/(httpredir|deb).debian.org/${APT_MIRROR:-deb.debian.org}/g" /etc/
+    --mount=type=cache,target=/root/.cache <<EOT
+  set -e
+  make
+  mkdir /out
+  mv ./criu/criu /out/
```
Perhaps use `/build/` as we do for all other stages

thaJeztah commented on Dockerfile:
```diff
@@ -16,15 +17,37 @@ RUN sed -ri "s/(httpredir|deb).debian.org/${APT_MIRROR:-deb.debian.org}/g" /etc/
 ENV GO111MODULE=off
 
+# criu
+FROM base AS criu-src
+WORKDIR /usr/src/criu
+RUN git init . && git remote add origin "https://github.com/checkpoint-restore/criu.git"
+ARG CRIU_VERSION
+RUN git fetch --depth 1 origin "${CRIU_VERSION}" && git checkout -q FETCH_HEAD
```
Curious why you're using a separate stage for this. Was this related to cross compiling (e.g.)? Otherwise, we may as well make it a `RUN` in the `criu` stage below

Would be nice if `ADD` could also extract, because then if would be just a

```dockerfile
ADD https://github.com/checkpoint-restore/criu/archive/refs/tags/${CRIU_VERSION}.tar.gz .
```

crazy-max replied:
Yes was related to cross comp with `TARGETPLATFORM` arg but not needed in this case.

crazy-max committed:
Dockerfile: pin criu version and build from source

pin criu for better reproducibility and build from source.
will solve issues we have with flakiness using the opensuse mirror
but also being able to build on other platforms

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
@tianon PTAL (sorry, looks like we're going back-and-forth on pre-built vs compile, but these package-repos are not very stable)

cpuguy83 commented:
:sadpanda:

I really don't want to have to build this from source every time. Yeah it gets cached, but I don't always have the build cached.
It would be really nice to add this as an image to ghcr or something that we can just reference in the Dockerfile.

neersighted commented:
I would like to see this (or some other approach) merged as it's bit me a few times -- can we reach consensus on this approach?

thaJeztah commented:
@cpuguy83 is that a "strong" no, or would you be ok with having this as an (intermediate?) solution?

@crazy-max also mentioned that the package is not available for armhf and riscv (but criu itself does support it? albeit still WIP); ideally that would be fixed in the SUSE repo of course.

While discussing; I was also looking where we actually use CRIU; I think this was for the integration tests, but it looks like the actual tests were broken, and currently skipped. For fun, I opened https://github.com/moby/moby/pull/44495 to see if they actually work.
